### PR TITLE
Refactor | RCE Web App

### PR DESCRIPTION
**Pull request by andrew-aiken**
**Overview of Change**
- Formatting
  - Jsonencode
  - `terraform fmt`
- Linting
  - default tags
  - variable types & descriptions
- LB log in S3 has "real" data
- S3 resource change
  - https://github.com/RhinoSecurityLabs/cloudgoat/issues/182
- Pull latest ubuntu ami with data call
  - Will work in regions other than `us-east-1`

**Testing**
Terraform version `1.2.0` & `1.5.6`

### CODE REVIEW

File: scenarios/rce_web_app/assets/elasticloadbalancing.log [hunk 1]
```diff
@@ -0,0 +1,32 @@
+http 2019-06-18T21:36:23.594569Z app/${cgid}/d36d4f13b73c2fe7 10.10.10.23:5132 10.0.10.254:9000 0.001 0.001 0.000 200 200 485 1287 "GET http://${load_balancer_dns}:80/mkja1xijqf0abo1h9glg.html HTTP/1.1" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.90 Safari/537.36" - - ${target_group_arn} "Root=1-5d095963-e2b838a764ed31d017b74cce" "-" "-" 0 2019-06-18T21:36:35.592000Z "forward" "-" "-"
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/assets/elasticloadbalancing.log:1]: Switch the default log file to use the values of the current deployed scenario

File: scenarios/rce_web_app/terraform/data_sources.tf [hunk 2]
```diff
@@ -1,8 +1,27 @@
 #AWS Account Id
-data "aws_caller_identity" "aws-account-id" {
-  
+data "aws_caller_identity" "aws-account-id" {}
+
+data "aws_ami" "ubuntu" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/data_sources.tf:4]: Use the latest ubuntu image, will work in multiple regions

File: scenarios/rce_web_app/terraform/data_sources.tf [hunk 3]
```diff
@@ -1,8 +1,27 @@
...
+    values = ["ebs"]
+  }
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
 }
-#S3 Full Access Policy
-data "aws_iam_policy" "s3-full-access" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/data_sources.tf:6]: Redundant

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 4]
```diff
@@ -1,146 +1,153 @@
...
-      "Action": "sts:AssumeRole",
-      "Principal": {
-        "Service": "ec2.amazonaws.com"
-      },
-      "Effect": "Allow",
-      "Sid": ""
+      Version = "2012-10-17"
+      Statement = [
+        {
+          Action = "sts:AssumeRole"
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:9]: jsonencode

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 5]
```diff
@@ -1,146 +1,153 @@
...
 }
-EOF
-  tags = {
-      Name = "cg-ec2-role-${var.cgid}"
-      Stack = "${var.stack-name}"
-      Scenario = "${var.scenario-name}"
-  }
-}
-#IAM Role Policy Attachment
-resource "aws_iam_role_policy_attachment" "cg-ec2-role-policy-attachment-s3" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:26]: Redundant

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 6]
```diff
@@ -1,146 +1,153 @@
...
+            Service = "ec2.amazonaws.com"
+          }
+          Effect = "Allow"
+          Sid    = ""
+        }
+      ]
     }
+  )
+
+  managed_policy_arns = [
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:20]: Attach policies directly on role creation

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 7]
```diff
@@ -1,146 +1,153 @@
...
     {
-      "Action": [
-        "ec2:DescribeInstances","ec2:DescribeVpcs", "ec2:DescribeSubnets",
-        "ec2:DescribeSecurityGroups", "rds:DescribeDBInstances"
-      ],
-      "Effect": "Allow",
-      "Resource": "*"
+      Version = "2012-10-17"
+      Statement = [
+        {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:38]: jsonencode

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 8]
```diff
@@ -1,146 +1,153 @@
...
+          Resource = "*"
+        }
+      ]
     }
-  ]
-}
-EOF
-}
-#IAM Role Policy Attachment
-resource "aws_iam_role_policy_attachment" "cg-ec2-role-policy-attachment-rds" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:51]: Redundant

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 9]
```diff
@@ -1,146 +1,153 @@
...
+  name        = "cg-ec2-ssh-${var.cgid}"
   description = "CloudGoat ${var.cgid} Security Group for EC2 Instance over SSH"
-  vpc_id = "${aws_vpc.cg-vpc.id}"
+  vpc_id      = aws_vpc.cg-vpc.id
   ingress {
-      from_port = 22
-      to_port = 22
-      protocol = "tcp"
-      cidr_blocks = var.cg_whitelist
+    from_port   = 22
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:70]: formatting

File: scenarios/rce_web_app/terraform/ec2.tf [hunk 10]
```diff
@@ -1,146 +1,153 @@
...
-        type = "ssh"
-        user = "ubuntu"
-        private_key = "${file(var.ssh-private-key-for-ec2)}"
-        host = self.public_ip
-      }
+  ami                         = data.aws_ami.ubuntu.image_id
+  instance_type               = "t2.micro"
+  iam_instance_profile        = aws_iam_instance_profile.cg-ec2-instance-profile.name
+  subnet_id                   = aws_subnet.cg-public-subnet-1.id
+  associate_public_ip_address = true
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/ec2.tf:125]: formatting

File: scenarios/rce_web_app/terraform/iam.tf [hunk 11]
```diff
@@ -1,121 +1,118 @@
 #IAM Users
 resource "aws_iam_user" "cg-lara" {
   name = "lara"
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/iam.tf:3]: Use default tags

File: scenarios/rce_web_app/terraform/iam.tf [hunk 12]
```diff
@@ -1,121 +1,118 @@
...
-      "Effect": "Allow",
-      "Resource": "*"
+      Version = "2012-10-17",
+      Statement = [
+        {
+          Action   = "s3:ListBucket"
+          Effect   = "Allow"
+          Resource = "arn:aws:s3:::cg-logs-s3-bucket-${local.cgid_suffix}"
+        },
+        {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/iam.tf:37]: jsonencode

File: scenarios/rce_web_app/terraform/iam.tf [hunk 13]
```diff
@@ -1,121 +1,118 @@
...
+          Action   = "s3:ListBucket"
+          Effect   = "Allow"
+          Resource = "arn:aws:s3:::cg-keystore-s3-bucket-${local.cgid_suffix}"
+        },
+        {
+          Action   = "s3:GetObject"
+          Effect   = "Allow"
+          Resource = "arn:aws:s3:::cg-keystore-s3-bucket-${local.cgid_suffix}/*"
+        },
+        {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/iam.tf:84]: jsonencode

File: scenarios/rce_web_app/terraform/lb.tf [hunk 14]
```diff
@@ -1,78 +1,76 @@
...
-      from_port = 80
-      to_port = 80
-      protocol = "tcp"
-      cidr_blocks = var.cg_whitelist
+    from_port   = 80
+    to_port     = 80
+    protocol    = "tcp"
+    cidr_blocks = var.cg_whitelist
   }
   egress {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/lb.tf:12]: formatting

File: scenarios/rce_web_app/terraform/locals.tf [hunk 15]
```diff
@@ -3,4 +3,9 @@ locals {
   # lowercase letters, numbers, and hyphens
   # Used for RDS, load balancer and S3
   cgid_suffix = replace(var.cgid, "/[^a-z0-9-]/", "-")
-}
\ No newline at end of file
+
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/locals.tf:6]: default tags for resources

File: scenarios/rce_web_app/terraform/rds.tf [hunk 16]
```diff
@@ -1,71 +1,68 @@
...
-      protocol = "tcp"
-      cidr_blocks = [
-          "10.0.10.0/24",
-          "10.0.20.0/24",
-          "10.0.30.0/24",
-          "10.0.40.0/24"
-      ]
+    from_port = 5432
+    to_port   = 5432
+    protocol  = "tcp"
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/rds.tf:9]: formatting

File: scenarios/rce_web_app/terraform/s3.tf [hunk 17]
```diff
@@ -1,115 +1,92 @@
...
+  })
 }
 
 #S3 Bucket Objects
-resource "aws_s3_bucket_object" "cg-lb-log-file" {
-  bucket = "${aws_s3_bucket.cg-logs-s3-bucket.id}"
-  key = "cg-lb-logs/AWSLogs/${data.aws_caller_identity.aws-account-id.account_id}/elasticloadbalancing/${var.region}/2019/06/19/555555555555_elasticloadbalancing_us-east-1_app.cg-lb-cgidp347lhz47g.d36d4f13b73c2fe7_20190618T2140Z_10.10.10.100_5m9btchz.log"
-  source = "../assets/555555555555_elasticloadbalancing_us-east-1_app.cg-lb-cgidp347lhz47g.d36d4f13b73c2fe7_20190618T2140Z_10.10.10.100_5m9btchz.log"
-  tags = {
+resource "aws_s3_object" "cg-lb-log-file" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/s3.tf:54]: Update to `aws_s3_object`

File: scenarios/rce_web_app/terraform/s3.tf [hunk 18]
```diff
@@ -1,115 +1,92 @@
...
-    Stack = "${var.stack-name}"
-    Scenario = "${var.scenario-name}"
-  }
-}
-
-resource "aws_s3_bucket_acl" "logs-s3-bucket-acl" {
-  bucket = aws_s3_bucket.cg-logs-s3-bucket.id
-  acl    = "private"
+  })
 }
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/s3.tf:75]: update to aws_s3_object

File: scenarios/rce_web_app/terraform/s3.tf [hunk 19]
```diff
@@ -1,115 +1,92 @@
...
+  bucket = aws_s3_bucket.cg-keystore-s3-bucket.id
+  key    = "cloudgoat"
+  source = var.ssh-private-key-for-ec2
+  tags = merge(local.default_tags, {
+    Name = "cg-ssh-private-key-file-${var.cgid}"
+  })
 }
 
-resource "aws_s3_bucket_acl" "keystore-s3-bucket-acl" {
+resource "aws_s3_object" "cg-ssh-public-key-file" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/s3.tf:85]: aws_s3_object

File: scenarios/rce_web_app/terraform/s3.tf [hunk 20]
```diff
@@ -1,115 +1,92 @@
...
+resource "aws_s3_object" "cg-ssh-private-key-file" {
+  bucket = aws_s3_bucket.cg-keystore-s3-bucket.id
+  key    = "cloudgoat"
+  source = var.ssh-private-key-for-ec2
+  tags = merge(local.default_tags, {
+    Name = "cg-ssh-private-key-file-${var.cgid}"
+  })
 }
 
-resource "aws_s3_bucket_acl" "keystore-s3-bucket-acl" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/s3.tf:112]: Remove bucket acl, default to be private

File: scenarios/rce_web_app/terraform/variables.tf [hunk 21]
```diff
@@ -1,41 +1,56 @@
-#Required: AWS Profile
 variable "profile" {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/variables.tf:1]: variable types & descriptions

File: scenarios/rce_web_app/terraform/versions.tf [hunk 22]
```diff
@@ -0,0 +1,10 @@
+terraform {
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/versions.tf:1]: Minimum versions

File: scenarios/rce_web_app/terraform/vpc.tf [hunk 23]
```diff
@@ -1,101 +1,96 @@
 #VPC
 resource "aws_vpc" "cg-vpc" {
-  cidr_block = "10.0.0.0/16"
+  cidr_block           = "10.0.0.0/16"
```
Comment (andrew-aiken) [inline scenarios/rce_web_app/terraform/vpc.tf:3]: Formatting

**Review by andrew-aiken — COMMENTED** [review]
Comments showing the overview of changes

**Review by jdearmas — APPROVED** [review]
I have reviewed and tested this code.

**Pull request merged**